README.md
# TweetWX - post weather updates to Twitter using OAuth

This set of scripts can be used to post status updates to your Twitter account for current weather conditions, weather forecasts, or advisories. It is currently set up to work with WeatherDisplay and the Saratoga templates, and the NOAA NWS forecast service.  
The script uses [OAuth](http://dev.twitter.com/pages/auth#intro) as all Twitter applications must use OAuth for authentication.

The OAuth process is more complicated than the old user:pass basic authentication. It's overall actions are shown in the diagram below:

<img src="./sample-oauth_diagram1.png" alt="OAUTH flow 1">

<img src="./sample-oauth_diagram2.png" alt="OAUTH flow 2">

This package has two setup functions that serve to obtain a durable access token/access token secret for the application to securely update your Twitter status with conditions from WD and forecasts from NOAA/NWS.  
My thanks to Joe Chung who wrote a simple OAuth/Twitter set of functions [documented on his blog](http://nullinfo.wordpress.com/oauth-twitter/) and these scripts are based on his code.

## Requirements:

*   These scripts need **PHP5+** to run. They will **not** run on PHP4.
*   You will need a valid [Twitter](https://twitter.com/) account to receive the postings.
*   These scripts assume you have the [WD/AJAX/PHP template](https://saratoga-weather.org/wxtemplates/index.php) running on your website being updated by [Weather-Display](https://www.weather-display.com/) software. (Conditions tweet is based on the _testtags.php_ data)
*   The TweetWX-forecast.php script uses the NOAA NWS forecast from [advforecast2.php](scripts-carterlake.php#advforecast) (WD-USA template set)
*   **Note:** if your site is running on **GoDaddy** linux hosting, you need to make a change in **TweetWX-global.php** in the function _do_post()_. Change the verify host and verify peer to FALSE (i.e.):  

    CURLOPT_SSL_VERIFYHOST => false,  
    CURLOPT_SSL_VERIFYPEER => false,  

    since their cURL does not currently support the verify host/peer options.  
    Thanks to Jason at [kswxlink.com](http://kswxlink.com/) for this discovery.

## Setup

1.  Download the distribution .zip file, unpack it preserving directory structure, and upload contents to your weather website.
2.  **(optional)** Register your copy of the application with Twitter at [http://twitter.com/oauth_clients/new](http://twitter.com/oauth_clients/new) using your Twitter account.  
    This is optional as you can use the built-in credentials for the application based on my Twitter application registration for TweetWX.  
    If you do your own registration, then copy the _consumer key_ and _consumer secret key_ values into the TweetWX-common.php values at the top of the page. Then save and upload the modified _TweetWX-common.php_ file.
3.  Run **TweetWX-setup1.php** on your website.  
    It will produce a page with two OAuth request values that must be copied into TweetWX-setup2.php and a link to Twitter to allow connection of the TweetWX app to your twitter account.  

    1.  Click on the link and Twitter will ask you to sign-in, then ask permission for TweetWX to be allowed access to your account.
    2.  If you agree, that will produce a Twitter page with a verification number, and that number needs to be copied into the proper variable inside of TweetWX-setup2.php.Update TweetWX-setup2.php with the two request keys and the verification number from Twitter and  
    upload the modified TweetWX-setup2.php to your website.
4.  Run **TweetWX-setup2.php** on your website. It will produce two keys as copy code.  
    These keys are to be copied into the top of _TweetWX-conditions.php_ and _TweetWX-forecast.php_ replacing the values there.  
    These OAuth credentials do not expire (AFAIK) and using them allows the script(s) to post to your Twitter account status.  
    **Keep these access keys secret** .. **they are similar to your user:pass on the old basic authentication scheme and allow read/write to your Twitter account.**
5.  Modify _TweetWX-conditions.php_ to insert the keys from your run of TweetWX-setup2.php (in step 4), and also change the directory pointer to where your _testtags.php_ is located (use relative FILE addressing).
6.  Modify _TweetWX-forecast.php_ to insert the keys from your run of _TweetWX-setup2.php_ (in step 4) and also change the directory point to where your _advforecast2.php_ script is located (relative file addressing)

If you want to redo your OAuth keys, then start at step (3) and proceed through step (6). I don't know how long the interim keys last, so make sure you do step (3) followed shortly by step (4). Once step (4) is completed, you can take as long as you like to do steps (5) and (6).

## Operation on your website

The best and most reliable method to execute the scripts on your website is via cron as a cronjob. Make sure you invoke PHP5 to run the scripts. I've found that a simple shell script makes running the cron job easier as you can set the home directory (etc).  
Something like cron-TweetWX-conditions.sh:

<div class="codebox">

```
#!/bin/bash
cd /my/home/directory/on/webserver/TweetWX
# Clean out old copies if any
ps auxw | grep "TweetWX-conditions.php" | grep -v 'grep' | awk '{print $2}' | xargs kill >/dev/null &2>1
/usr/bin/php5 -q TweetWX-conditions.php > TweetWX-conditions.tmp
cp TweetWX-conditions.tmp TweetWX-conditions.html
#
```

Which will run the script and save the last output into TweetWX-conditions.html so you can follow the progress of the last execution.  
Don't forget to _chmod +x_ the script above (after changing the _cd_ statement for your website document root). A crontab entry of
```
4 * * * *  /my/home/directory/on/webserver/cron-TweetWX-conditions.sh</pre>
```

works to tweet my conditions at 4 minutes after the hour, 24 times a day. I use a similar script for the forecasts, but the schedule is for execution by cron as 47 past the hour at 3,6,10,15,18 and 22h:

```
47 3,6,10,15,18,22 * * *  /my/home/directory/on/webserver/cron-TweetWX-forecast.sh</pre>
```

If you don't happen to have cron available on your webhoster, then you can configure Weather-Display to do HTML downloads on the TweetWX-conditions.php and TweetWX-forecast.php webpages from your website. Use WD, Control Panel, FTP & Connections, HTTP Download TAB. You have 3 setups you can use, so you can have one run hourly for your conditions and a different one run every 4 hours for your forecasts.


CHANGELOG.txt
TweetWX package changes
---

1.0.0  2010-04-14  Joe Chung
* Initial Release

1.0.0  2010-06-11  Ken True
* initial release of TweetWX package incorporating Joe's OAuth Examples using Twitter for use with
* Weather-Display, WD/AJAX/PHP template websites

1.01   2013-06-13  Ken True
* adapted TweetWX-globals.php for Twitter API V1.1 using SSL

1.02   2014-01-18  Ken True
* fixes for Twitter OAuth setup using SSL only

1.03   27-May-2014 Ken True
* fixed TweetWX-setup1.php Twitter link for $oauth_verifier to use https/SSL


TweetWX-setup1.php
<?php
/* TweetWX-setup1.php

 This package is based on code by Joe Chung at http://nullinfo.wordpress.com/oauth-twitter/
 
 Repackaged for TweetWX by Ken True - Saratoga-weather.org  08-Jun-2010
 Docs at http://saratoga-weather.org/scripts-TweetWX.php
 TweetWX-setup1.php  V1.00 - 11-Jun-2010 - Initial relase

 This utility function should be used once to get the OAuth request_token and request_token_secret
 for use in the next step of setup.
 
// Version 1.01 - 12-Jul-2013 - added debugging output if request fails for token
// Version 1.02 - 18-Jan-2014 - fixes for OAuth using SSL only
// Version 1.03 - 27-May-2014 - fixed Twitter link for $oauth_verifier to use https/SSL
 
*/
require 'TweetWX-globals.php';

if (phpversion() < 5) {
  echo 'Failure: This Script requires PHP version 5 or greater. You only have PHP version: ' . phpversion();
  exit;
}

?>
<h1>TweetWX Setup - part 1</h1>
<p>This step of the setup will obtain a request token and request token secret string for use by
TweetWX to connect to your twitter account.  It has two parts:
<ul>
  <li>Get the request token and request secret token for use by the app in the future (values shown below)</li>
  <li>You must click on the hotlink below to go to Twitter, authorize this app access to your account, and
  copy the verification number off the Twitter authorization screen for use in the $oauth_verifier variable</li>
</ul>

<?php
// Callback can either be 'oob' or a url
$callback='oob';

// Get the request token using HTTP GET and HMAC-SHA1 signature
$retarr = get_request_token(OAUTH_CONSUMER_KEY, OAUTH_CONSUMER_SECRET,
                            $callback, true, true, true);

if (! empty($retarr)) {
  list($info, $headers, $body, $body_parsed) = $retarr;
  if ($info['http_code'] == 200 && !empty($body)) {
    print "\n<p>Click on <a href=\"" .
        "https://api.twitter.com/oauth/authorize?" .
        rfc3986_decode($body) . "\" target=\"_blank\"><strong>this link</strong></a>
		to get the verification code for Step2 in setup</p>\n";
	  print "<pre>\n";
	  print "Copy the following two lines of PHP assignment statements into TweetWX-setup2.php to replace the two values in the script\n";
      print "\n#--copy code below-------------\n";
	  print '$request_token = \''.$body_parsed['oauth_token'].'\'; // oauth_token'."\n";
	  print '$request_token_secret = \''.$body_parsed['oauth_token_secret'].'\'; // oauth_token_secret'."\n";
	  print "#----end copy code above---------\n\n";
	  print "Then copy the verification number from the Twitter authorization script into the \$oauth_verifier variable\n";
	  print "</pre>\n";
?>	
<p>After you have updated TweetWX-setup2.php with the three values from above, and uploaded the updated copy,
then proceed to <a href="TweetWX-setup2.php"><strong>TweetWX-setup2.php</strong></a> to get the final OAuth
tokens to use with TweetWX and your twitter account.</p>
<?php
  } else {
	  print "<p>Something did not work right... here is the return from the request.</p>\n";
	  print "<pre>\n";
	  print "RC=".$info['http_code']."\n";
	  print "----- info array -----\n";
	  print_r($info,true);
	  print "-----headers of response-----\n";
	  print "$headers\n";
	  print "-----body of response------\n";
	  print "$body\n";
	  print "</pre>\n";
	
  }
}

exit(0);

/**
 * Get a request token.
 * @param string $consumer_key obtained when you registered your app
 * @param string $consumer_secret obtained when you registered your app
 * @param string $callback callback url can be the string 'oob'
 * @param bool $usePost use HTTP POST instead of GET
 * @param bool $useHmacSha1Sig use HMAC-SHA1 signature
 * @param bool $passOAuthInHeader pass OAuth credentials in HTTP header
 * @return array of response parameters or empty array on error
 */
function get_request_token($consumer_key, $consumer_secret, $callback, $usePost=false, $useHmacSha1Sig=true, $passOAuthInHeader=false)
{
  $retarr = array();  // return value
  $response = array();

  $url = 'https://api.twitter.com/oauth/request_token';
  $params['oauth_version'] = '1.0';
  $params['oauth_nonce'] = mt_rand();
  $params['oauth_timestamp'] = time();
  $params['oauth_consumer_key'] = $consumer_key;
  $params['oauth_callback'] = $callback;

  // compute signature and add it to the params list
  if ($useHmacSha1Sig) {
    $params['oauth_signature_method'] = 'HMAC-SHA1';
    $params['oauth_signature'] =
      oauth_compute_hmac_sig($usePost? 'POST' : 'GET', $url, $params,
                             $consumer_secret, null);
  } else {
    $params['oauth_signature_method'] = 'PLAINTEXT';
    $params['oauth_signature'] =
      oauth_compute_plaintext_sig($consumer_secret, null);
  }

  // Pass OAuth credentials in a separate header or in the query string
  if ($passOAuthInHeader) {
    $query_parameter_string = oauth_http_build_query($params, true);
    $header = build_oauth_header($params, "Twitter API");
    $headers[] = $header;
  } else {
    $query_parameter_string = oauth_http_build_query($params);
  }

  // POST or GET the request
  if ($usePost) {
    $request_url = $url;
    logit("getreqtok:INFO:request_url:$request_url");
    logit("getreqtok:INFO:post_body:$query_parameter_string");
    $headers[] = 'Content-Type: application/x-www-form-urlencoded';
    $response = do_post($request_url, $query_parameter_string, 443, $headers);
  } else {
    $request_url = $url . ($query_parameter_string ?
                           ('?' . $query_parameter_string) : '' );
    logit("getreqtok:INFO:request_url:$request_url");
    $response = do_get($request_url, 443, $headers);
  }

  // extract successful response
  if (! empty($response)) {
    list($info, $header, $body) = $response;
    $body_parsed = oauth_parse_str($body);
    if (! empty($body_parsed)) {
      logit("getreqtok:INFO:response_body_parsed:");
    }
    $retarr = $response;
    $retarr[] = $body_parsed;
  }

  return $retarr;
}
?>


TweetWX-setup2.php
<?php
/* TweetWX-setup2.php

 This package is based on code by Joe Chung at http://nullinfo.wordpress.com/oauth-twitter/
 
 Repackaged for TweetWX by Ken True - Saratoga-weather.org  08-Jun-2010
 Docs at http://saratoga-weather.org/scripts-TweetWX.php
 TweetWX-setup2.php  V1.00 - 11-Jun-2010 - Initial relase

 This utility function should be used once to get the OAuth access_token and access_token_secret
 for use with the TweetWX-forecast.php and TweetWX-conditions.php.  It should only need to be run once.

// Version 1.01 - 12-Jul-2013 - added debugging output if request fails for token
// Version 1.02 - 18-Jan-2014 - fixes for OAuth using SSL only
 
*/
require 'TweetWX-globals.php';

// Fill in the next 3 variables from data by running TweetWX-setup1.php:
//
$request_token = '-replace-with-token-from-TweetWX-setup1-'; // oauth_token
$request_token_secret = '-replace-with-token-secret-from-TweetWX-setup1-'; // oauth_token_secret
$oauth_verifier= '-replace-with-verification-code-from-twitter';
//
if (phpversion() < 5) {
  echo 'Failure: This Script requires PHP version 5 or greater. You only have PHP version: ' . phpversion();
  exit;
}
//
?>
<h1>TweetWX Setup - part 2</h1>
<p>This last step of the setup will obtain a request token and request token secret string for use by
TweetWX to connect to your twitter account.</p>

<?php


// Get the access token using HTTP GET and HMAC-SHA1 signature
$retarr = get_access_token(OAUTH_CONSUMER_KEY, OAUTH_CONSUMER_SECRET,
                           $request_token, $request_token_secret,
                           $oauth_verifier, true, true, true);
if (! empty($retarr)) {
  list($info, $headers, $body, $body_parsed) = $retarr;
  if ($info['http_code'] == 200 && !empty($body)) {
	  print "<pre>\n";
	  print "Copy the following two lines of PHP assignment statements into TweetWX-conditions.php and TweetWX-forecast.php to replace the two values in the script.\n";
      print "\n#--copy code below-------------\n";
	  print '$access_token = \''.$body_parsed['oauth_token'].'\'; // oauth_token'."\n";
	  print '$access_token_secret = \''.$body_parsed['oauth_token_secret'].'\'; // oauth_token_secret'."\n";
	  print "#----end copy code above---------\n\n";
	  
	  print "Your setup of TweetWX is now complete.  The two access codes above will enable the scripts to\n";
	  print "securely connect and update your twitter account with status updates for conditions and forecasts.\n";
	  print "</pre>\n";
  } else {
	  print "<p>Something did not work right... here is the return from the request.</p>\n";
	  print "<pre>\n";
	  print "RC=".$info['http_code']."\n";
	  print "----- info array -----\n";
	  print_r($info,true);
	  print "-----headers of response-----\n";
	  print "$headers\n";
	  print "-----body of response------\n";
	  print "$body\n";
	  print "</pre>\n";
  }
}

exit(0);

/**
 * Get an access token using a request token and OAuth Verifier.
 * @param string $consumer_key obtained when you registered your app
 * @param string $consumer_secret obtained when you registered your app
 * @param string $request_token obtained from getreqtok
 * @param string $request_token_secret obtained from getreqtok
 * @param string $oauth_verifier obtained from twitter oauth/authorize
 * @param bool $usePost use HTTP POST instead of GET
 * @param bool $useHmacSha1Sig use HMAC-SHA1 signature
 * @param bool $passOAuthInHeader pass OAuth credentials in HTTP header
 * @return array of response parameters or empty array on error
 */
function get_access_token($consumer_key, $consumer_secret, $request_token, $request_token_secret, $oauth_verifier, $usePost=false, $useHmacSha1Sig=true, $passOAuthInHeader=true)
{
  $retarr = array();  // return value
  $response = array();

  $url = 'https://api.twitter.com/oauth/access_token';
  $params['oauth_version'] = '1.0';
  $params['oauth_nonce'] = mt_rand();
  $params['oauth_timestamp'] = time();
  $params['oauth_consumer_key'] = $consumer_key;
  $params['oauth_token']= $request_token;
  $params['oauth_verifier'] = $oauth_verifier;

  // compute signature and add it to the params list
  if ($useHmacSha1Sig) {
    $params['oauth_signature_method'] = 'HMAC-SHA1';
    $params['oauth_signature'] =
      oauth_compute_hmac_sig($usePost? 'POST' : 'GET', $url, $params,
                             $consumer_secret, $request_token_secret);
  } else {
    $params['oauth_signature_method'] = 'PLAINTEXT';
    $params['oauth_signature'] =
      oauth_compute_plaintext_sig($consumer_secret, $request_token_secret);
  }

  // Pass OAuth credentials in a separate header or in the query string
  if ($passOAuthInHeader) {
    $query_parameter_string = oauth_http_build_query($params, true);
    $header = build_oauth_header($params, "Twitter API");
    $headers[] = $header;
  } else {
    $query_parameter_string = oauth_http_build_query($params);
  }

  // POST or GET the request
  if ($usePost) {
    $request_url = $url;
    logit("getacctok:INFO:request_url:$request_url");
    logit("getacctok:INFO:post_body:$query_parameter_string");
    $headers[] = 'Content-Type: application/x-www-form-urlencoded';
    $response = do_post($request_url, $query_parameter_string, 443, $headers);
  } else {
    $request_url = $url . ($query_parameter_string ?
                           ('?' . $query_parameter_string) : '' );
    logit("getacctok:INFO:request_url:$request_url");
    $response = do_get($request_url, 443, $headers);
  }

  // extract successful response
  if (! empty($response)) {
    list($info, $header, $body) = $response;
    $body_parsed = oauth_parse_str($body);
    if (! empty($body_parsed)) {
      logit("getacctok:INFO:response_body_parsed:");
    }
    $retarr = $response;
    $retarr[] = $body_parsed;
  }

  return $retarr;
}
?>


TweetWX-globals.php
<?php
/**
 * Globals and utilities for OAuth Examaples package
 This package is based on code by Joe Chung at http://nullinfo.wordpress.com/oauth-twitter/
 
 Repackaged for TweetWX by Ken True - Saratoga-weather.org  08-Jun-2010
 Docs at http://saratoga-weather.org/scripts-TweetWX.php
 Version 1.00 - 11-Jun-2010 - Initial relase
 Version 1.01 - 13-Jun-2013 - fixes for API 1.1 using SSL only
 Version 1.02 - 18-Jan-2014 - fixes for OAuth using SSL only
 
 */

// you can leave these two settings alone if you like.  They are for TweetWX script from SaratogaWX
//
// If you choose to define your own app using http://twitter.com/oauth_clients/new then
// replace the two values below with the values given by twitter for the customer_key and customer_secret.
define('OAUTH_CONSUMER_KEY', 'fJkVCHCtIT0drcp18Yr7Ug');
define('OAUTH_CONSUMER_SECRET', 'DsdVaBnI9jeF39tN5PQPdZeV0QbkA70os74HbzL5Ws');
//
//
// ------------ No changes are needed below ----------------------------------
$progname = $argv[0];
$debug = 0; // Set to 1 for verbose debugging output BUT DON'T DO IN PRODUCTION as your keys will be visible

if (phpversion() < 5) {
  echo 'Failure: This Script requires PHP version 5 or greater. You only have PHP version: ' . phpversion();
  exit;
}

function logit($msg,$preamble=true)
{
  //  date_default_timezone_set('America/Los_Angeles');
  global $debug;
  if(! $debug) { return; }
  $now = date('Ymd-His', time());
  //error_log(($preamble ? "+++${now}:" : '') . $msg);
  print "<!-- log: $now $msg -->\n";
}

/**
 * Do an HTTP GET
 * @param string $url
 * @param int $port (optional)
 * @param array $headers an array of HTTP headers (optional)
 * @return array ($info, $header, $response) on success or empty array on error.
 */
function do_get($url, $port=443, $headers=NULL)
{
  $retarr = array();  // Return value

  $curl_opts = array(CURLOPT_URL => $url,
                     CURLOPT_PORT => $port,
                     CURLOPT_POST => false,
                     CURLOPT_SSL_VERIFYHOST => false,
                     CURLOPT_SSL_VERIFYPEER => false,
                     CURLOPT_RETURNTRANSFER => true);

  if ($headers) { $curl_opts[CURLOPT_HTTPHEADER] = $headers; }

  $response = do_curl($curl_opts);

  if (! empty($response)) { $retarr = $response; }

  return $retarr;
}

/**
 * Do an HTTP POST
 * @param string $url
 * @param int $port (optional)
 * @param array $headers an array of HTTP headers (optional)
 * @return array ($info, $header, $response) on success or empty array on error.
 */
function do_post($url, $postbody, $port=443, $headers=NULL)
{
  $retarr = array();  // Return value

  $curl_opts = array(CURLOPT_URL => $url,
                     CURLOPT_PORT => $port,
                     CURLOPT_POST => true,
                     CURLOPT_SSL_VERIFYHOST => 2,
                     CURLOPT_SSL_VERIFYPEER => true,
                     CURLOPT_POSTFIELDS => $postbody,
                     CURLOPT_RETURNTRANSFER => true);

  if ($headers) { $curl_opts[CURLOPT_HTTPHEADER] = $headers; }

  $response = do_curl($curl_opts);

  if (! empty($response)) { $retarr = $response; }

  return $retarr;
}

/**
 * Make a curl call with given options.
 * @param array $curl_opts an array of options to curl
 * @return array ($info, $header, $response) on success or empty array on error.
 */
function do_curl($curl_opts)
{
  global $debug;

  $retarr = array();  // Return value

  if (! $curl_opts) {
    if ($debug) { logit("do_curl:ERR:curl_opts is empty"); }
    return $retarr;
  }

  // Open curl session
  $ch = curl_init();
  if (! $ch) {
    if ($debug) { logit("do_curl:ERR:curl_init failed"); }
    return $retarr;
  }

  // Set curl options that were passed in
  curl_setopt_array($ch, $curl_opts);

  // Ensure that we receive full header
  curl_setopt($ch, CURLOPT_HEADER, true);

  if ($debug) {
    curl_setopt($ch, CURLINFO_HEADER_OUT, true);
    curl_setopt($ch, CURLOPT_VERBOSE, true);
  }

  // Send the request and get the response
  ob_start();
  $response = curl_exec($ch);
  $curl_spew = ob_get_contents();
  ob_end_clean();
  if ($debug && $curl_spew) {
    logit("do_curl:INFO:curl_spew begin");
    logit($curl_spew, false);
    logit("do_curl:INFO:curl_spew end");
  }

  // Check for errors
  if (curl_errno($ch)) {
    $errno = curl_errno($ch);
    $errmsg = curl_error($ch);
    if ($debug) { logit("do_curl:ERR:$errno:$errmsg"); }
    curl_close($ch);
    unset($ch);
    return $retarr;
  }

  if ($debug) {
    logit("do_curl:DBG:header sent begin");
    $header_sent = curl_getinfo($ch, CURLINFO_HEADER_OUT);
    logit($header_sent, false);
    logit("do_curl:DBG:header sent end");
  }

  // Get information about the transfer
  $info = curl_getinfo($ch);

  // Parse out header and body
  $header_size = curl_getinfo($ch, CURLINFO_HEADER_SIZE);
  $header = substr($response, 0, $header_size);
  $body = substr($response, $header_size );

  // Close curl session
  curl_close($ch);
  unset($ch);

  if ($debug) {
    logit("do_curl:DBG:response received begin");
    if (!empty($response)) { logit($response, false); }
    logit("do_curl:DBG:response received end");
  }

  // Set return value
  array_push($retarr, $info, $header, $body);

  return $retarr;
}

/**
 * Pretty print some JSON
 * @param string $json The packed JSON as a string
 * @param bool $html_output true if the output should be escaped
 * (for use in HTML)
 * @link http://us2.php.net/manual/en/function.json-encode.php#80339
 */
function json_pretty_print($json, $html_output=false)
{
  $spacer = '  ';
  $level = 1;
  $indent = 0; // current indentation level
  $pretty_json = '';
  $in_string = false;

  $len = strlen($json);

  for ($c = 0; $c < $len; $c++) {
    $char = $json[$c];
    switch ($char) {
    case '{':
    case '[':
      if (!$in_string) {
        $indent += $level;
        $pretty_json .= $char . "\n" . str_repeat($spacer, $indent);
      } else {
        $pretty_json .= $char;
      }
      break;
    case '}':
    case ']':
      if (!$in_string) {
        $indent -= $level;
        $pretty_json .= "\n" . str_repeat($spacer, $indent) . $char;
      } else {
        $pretty_json .= $char;
      }
      break;
    case ',':
      if (!$in_string) {
        $pretty_json .= ",\n" . str_repeat($spacer, $indent);
      } else {
        $pretty_json .= $char;
      }
      break;
    case ':':
      if (!$in_string) {
        $pretty_json .= ": ";
      } else {
        $pretty_json .= $char;
      }
      break;
    case '"':
      if ($c > 0 && $json[$c-1] != '\\') {
        $in_string = !$in_string;
      }
    default:
      $pretty_json .= $char;
      break;
    }
  }

  return ($html_output) ?
    '<pre>' . htmlentities($pretty_json) . '</pre>' :
    $pretty_json . "\n";
}

/**
 * Call twitter to post a tweet
 * @param string $consumer_key obtained when you registered your app
 * @param string $consumer_secret obtained when you registered your app
 * @param string $status_message
 * @param string $access_token obtained from get_request_token
 * @param string $access_token_secret obtained from get_request_token
 * @param bool $usePost use HTTP POST instead of GET
 * @param bool $passOAuthInHeader pass OAuth credentials in HTTP header
 * @return response string or empty array on error
 */
function post_tweet($consumer_key, $consumer_secret, $status_message, $access_token, $access_token_secret, $usePost=true, $passOAuthInHeader=true)
{
  $retarr = array();  // return value
  $response = array();

//  $url = 'http://api.twitter.com/1/statuses/update.json';
  $url = 'https://api.twitter.com/1.1/statuses/update.json';
  $params['status'] = $status_message;
  $params['oauth_version'] = '1.0';
  $params['oauth_nonce'] = mt_rand();
  $params['oauth_timestamp'] = time();
  $params['oauth_consumer_key'] = $consumer_key;
  $params['oauth_token'] = $access_token;

  // compute hmac-sha1 signature and add it to the params list
  $params['oauth_signature_method'] = 'HMAC-SHA1';
  $params['oauth_signature'] =
      oauth_compute_hmac_sig($usePost? 'POST' : 'GET', $url, $params,
                             $consumer_secret, $access_token_secret);

  // Pass OAuth credentials in a separate header or in the query string
  if ($passOAuthInHeader) {
    $query_parameter_string = oauth_http_build_query($params, true);
    $header = build_oauth_header($params, "Twitter API");
    $headers[] = $header;
  } else {
    $query_parameter_string = oauth_http_build_query($params);
  }

  // POST or GET the request
  if ($usePost) {
    $request_url = $url;
    logit("tweet:INFO:request_url:$request_url");
    logit("tweet:INFO:post_body:$query_parameter_string");
    $headers[] = 'Content-Type: application/x-www-form-urlencoded';
    $response = do_post($request_url, $query_parameter_string, 443, $headers);
  } else {
    $request_url = $url . ($query_parameter_string ?
                           ('?' . $query_parameter_string) : '' );
    logit("tweet:INFO:request_url:$request_url");
    $response = do_get($request_url, 443, $headers);
  }

  // extract successful response
  if (! empty($response)) {
    list($info, $header, $body) = $response;
    if ($body) {
      logit("tweet:INFO:response:");
      print(json_pretty_print($body));
    }
    $retarr = $response;
  }

  return $retarr;
}
/**
 * Helper functions for OAuth
 */

/**
 * Build a query parameter string according to OAuth Spec.
 * @param array $params an array of query parameters
 * @return string all the query parameters properly sorted and encoded
 * according to the OAuth spec, or an empty string if params is empty.
 * @link http://oauth.net/core/1.0/#rfc.section.9.1.1
 */
function oauth_http_build_query($params, $excludeOauthParams=false)
{
  $query_string = '';
  if (! empty($params)) {

    // rfc3986 encode both keys and values
    $keys = rfc3986_encode(array_keys($params));
    $values = rfc3986_encode(array_values($params));
    $params = array_combine($keys, $values);

    // Parameters are sorted by name, using lexicographical byte value ordering.
    // http://oauth.net/core/1.0/#rfc.section.9.1.1
    uksort($params, 'strcmp');

    // Turn params array into an array of "key=value" strings
    $kvpairs = array();
    foreach ($params as $k => $v) {
      if ($excludeOauthParams && substr($k, 0, 5) == 'oauth') {
        continue;
      }
      if (is_array($v)) {
        // If two or more parameters share the same name,
        // they are sorted by their value. OAuth Spec: 9.1.1 (1)
        natsort($v);
        foreach ($v as $value_for_same_key) {
          array_push($kvpairs, ($k . '=' . $value_for_same_key));
        }
      } else {
        // For each parameter, the name is separated from the corresponding
        // value by an '=' character (ASCII code 61). OAuth Spec: 9.1.1 (2)
        array_push($kvpairs, ($k . '=' . $v));
      }
    }

    // Each name-value pair is separated by an '&' character, ASCII code 38.
    // OAuth Spec: 9.1.1 (2)
    $query_string = implode('&', $kvpairs);
  }

  return $query_string;
}

/**
 * Parse a query string into an array.
 * @param string $query_string an OAuth query parameter string
 * @return array an array of query parameters
 * @link http://oauth.net/core/1.0/#rfc.section.9.1.1
 */
function oauth_parse_str($query_string)
{
  $query_array = array();

  if (isset($query_string)) {

    // Separate single string into an array of "key=value" strings
    $kvpairs = explode('&', $query_string);

    // Separate each "key=value" string into an array[key] = value
    foreach ($kvpairs as $pair) {
      list($k, $v) = explode('=', $pair, 2);

      // Handle the case where multiple values map to the same key
      // by pulling those values into an array themselves
      if (isset($query_array[$k])) {
        // If the existing value is a scalar, turn it into an array
        if (is_scalar($query_array[$k])) {
          $query_array[$k] = array($query_array[$k]);
        }
        array_push($query_array[$k], $v);
      } else {
        $query_array[$k] = $v;
      }
    }
  }

  return $query_array;
}

/**
 * Build an OAuth header for API calls
 * @param array $params an array of query parameters
 * @return string encoded for insertion into HTTP header of API call
 */
function build_oauth_header($params, $realm='')
{
  $header = 'Authorization: OAuth realm="' . $realm . '"';
  foreach ($params as $k => $v) {
    if (substr($k, 0, 5) == 'oauth') {
      $header .= ',' . rfc3986_encode($k) . '="' . rfc3986_encode($v) . '"';
    }
  }
  return $header;
}

/**
 * Compute an OAuth PLAINTEXT signature
 * @param string $consumer_secret
 * @param string $token_secret
 */
function oauth_compute_plaintext_sig($consumer_secret, $token_secret)
{
  return ($consumer_secret . '&' . $token_secret);
}

/**
 * Compute an OAuth HMAC-SHA1 signature
 * @param string $http_method GET, POST, etc.
 * @param string $url
 * @param array $params an array of query parameters for the request
 * @param string $consumer_secret
 * @param string $token_secret
 * @return string a base64_encoded hmac-sha1 signature
 * @see http://oauth.net/core/1.0/#rfc.section.A.5.1
 */
function oauth_compute_hmac_sig($http_method, $url, $params, $consumer_secret, $token_secret)
{
  global $debug;

  $base_string = signature_base_string($http_method, $url, $params);
  $signature_key = rfc3986_encode($consumer_secret) . '&' . rfc3986_encode($token_secret);
  $sig = base64_encode(hash_hmac('sha1', $base_string, $signature_key, true));
  if ($debug) {
    logit("oauth_compute_hmac_sig:DBG:sig:$sig");
  }
  return $sig;
}

/**
 * Make the URL conform to the format scheme://host/path
 * @param string $url
 * @return string the url in the form of scheme://host/path
 */
function normalize_url($url)
{
  $parts = parse_url($url);

  $scheme = $parts['scheme'];
  $host = $parts['host'];
  $port = '';
  if(isset($parts['port'])) {$port = $parts['port'];}
  $path = $parts['path'];

  if (! $port) {
    $port = ($scheme == 'https') ? '443' : '80';
  }
  if (($scheme == 'https' && $port != '443')
      || ($scheme == 'http' && $port != '80')) {
    $host = "$host:$port";
  }

  return "$scheme://$host$path";
}

/**
 * Returns the normalized signature base string of this request
 * @param string $http_method
 * @param string $url
 * @param array $params
 * The base string is defined as the method, the url and the
 * parameters (normalized), each urlencoded and the concated with &.
 * @see http://oauth.net/core/1.0/#rfc.section.A.5.1
 */
function signature_base_string($http_method, $url, $params)
{
  // Decompose and pull query params out of the url
  $query_str = parse_url($url, PHP_URL_QUERY);
  if ($query_str) {
    $parsed_query = oauth_parse_str($query_str);
    // merge params from the url with params array from caller
    $params = array_merge($params, $parsed_query);
  }

  // Remove oauth_signature from params array if present
  if (isset($params['oauth_signature'])) {
    unset($params['oauth_signature']);
  }

  // Create the signature base string. Yes, the $params are double encoded.
  $base_string = rfc3986_encode(strtoupper($http_method)) . '&' .
                 rfc3986_encode(normalize_url($url)) . '&' .
                 rfc3986_encode(oauth_http_build_query($params));

  logit("signature_base_string:INFO:normalized_base_string:$base_string");

  return $base_string;
}

/**
 * Encode input per RFC 3986
 * @param string|array $raw_input
 * @return string|array properly rfc3986 encoded raw_input
 * If an array is passed in, rfc3896 encode all elements of the array.
 * @link http://oauth.net/core/1.0/#encoding_parameters
 */
function rfc3986_encode($raw_input)
{
  if (is_array($raw_input)) {
    return array_map('rfc3986_encode', $raw_input);
  } else if (is_scalar($raw_input)) {
    return str_replace('%7E', '~', rawurlencode($raw_input));
  } else {
    return '';
  }
}

function rfc3986_decode($raw_input)
{
  return rawurldecode($raw_input);
}

?>
